packetbeat/protos/smtp/smtp.go

```diff
@@ -0,0 +1,212 @@
+type connection struct {
+	streams [2]*stream
+	trans   transactions
+
+	// Packet loss affects SMTP session parsing
+	isPacketLoss bool
+}
+
+// Uni-directioal tcp stream state for parsing messages.
+type stream struct {
+	parser parser
+}
+
+type parseState uint8
+
+const (
+	stateCommand parseState = iota
+	// Request only state
+	stateData
+)
+
+var (
+	debugf = logp.MakeDebug("smtp")
+
+	// use isDebug to guard debugf to minimize allocations (garbage
+	// collection) when debug log is disabled.
+	isDebug = false
+)
+
+func init() {
+	protos.Register("smtp", New)
+}
+
+// New create and initializes a new smtp protocol analyzer instance.
+func New(
+	testMode bool,
+	results protos.Reporter,
+	cfg *common.Config,
+) (protos.Plugin, error) {
+	p := &smtpPlugin{}
+	config := defaultConfig
+	if !testMode {
+		if err := cfg.Unpack(&config); err != nil {
+			return nil, err
+		}
+	}
+
+	if err := p.init(results, &config); err != nil {
+		return nil, err
+	}
+	return p, nil
+}
+
+func (smtp *smtpPlugin) init(results protos.Reporter, config *smtpConfig) error {
+	if err := smtp.setFromConfig(config); err != nil {
+		return err
+	}
+	smtp.pub.results = results
+
+	isDebug = logp.IsDebug("smtp")
+	return nil
+}
+
+func (smtp *smtpPlugin) setFromConfig(config *smtpConfig) error {
+
+	// set module configuration
+	if err := smtp.ports.Set(config.Ports); err != nil {
+		return err
+	}
+
+	// set parser configuration
+	parser := &smtp.parserConfig
+	parser.maxBytes = tcp.TCPMaxDataInStream
+
+	// set transaction correlator configuration
+	trans := &smtp.transConfig
+	trans.transactionTimeout = config.TransactionTimeout
+
+	// set transaction publisher configuration
+	pub := &smtp.pub
+	pub.sendRequest = config.SendRequest
+	pub.sendResponse = config.SendResponse
+	pub.sendDataHeaders = config.SendDataHeaders
+	pub.sendDataBody = config.SendDataBody
+
+	return nil
+}
+
+// ConnectionTimeout returns the per stream connection timeout.
+// Return <=0 to set default tcp module transaction timeout.
+func (smtp *smtpPlugin) ConnectionTimeout() time.Duration {
+	return smtp.transConfig.transactionTimeout
+}
+
+// GetPorts returns the ports numbers packets shall be processed for.
+func (smtp *smtpPlugin) GetPorts() []int {
+	return smtp.ports.Ports
+}
+
+// Parse processes a TCP payload. Returns nil if connection state
+// should be dropped (e.g. parser not in sync with tcp stream)
+func (smtp *smtpPlugin) Parse(
+	pkt *protos.Packet,
+	tcptuple *common.TCPTuple,
+	dir uint8,
+	private protos.ProtocolData,
+) protos.ProtocolData {
+	defer logp.Recover("Parse SMTP exception")
+
+	conn := smtp.ensureConnection(private)
+
+	// There's no way to recover synchronization from packet loss
+	// during SMTP session. However packetbeat doesn't let us simply
+	// drop the session. The best we can do is suppress any message
+	// generation for the rest of the session.
+	if conn.isPacketLoss {
+		return conn
```

Comment:
Problem exists even on startup, as packetbeat might be started in the middle of an SMTP session. The parser should be hardened enough to return errors + ensure state is dropped if it expects to parse an SMTP header. At some point packetbeat needs to sync with an already active connection (by luck). This is true for the other protocols as well. We just try to parse and drop state, until we're finally/eventually in sync with an already existing connection.

I see how SMTP can be somewhat ambigious when not being in sync. One option trying to sync is looking for specific message header/type at beginning of line or look out for final `<CRLF>.<CRLF>` sequence to start syncing after current message has been transferred in case a session is used to send multiple mails.

Reply:
Great idea, `<CRLF>.<CRLF>` can only happen inside of a `DATA` request, so it's a good way to re-sync the parser.